File: Python/Contest 4/2021 - 2022/FibonacciMandelbrotIntermediate.py
#
# ACSL 2021-2022 - Contest 4 - Fibonacci & Mandelbrot - Intermediate Division
#
#

#
# Complete the 'cycleLength' function below.
#
# The function is expected to return a STRING.
# The function accepts following parameters:
#  1. FLOAT realPartC
#  2. FLOAT imagPartC
#
def cycle_length(real_part_c, imag_part_c):

    c = [real_part_c, imag_part_c]
    number = [0.0, 0.0]
    number_list = [number]
    n = 0

    while n <= 500:
        n += 1
        number = function(number, c)

        if absolute_value_square(number) > 16:
            return "ESCAPES " + str(n)

        for k in range(n):
            if number_list[k][0] == number[0] and number_list[k][1] == number[1]:
                return str(n - k)

        number_list.append(number)

    return ""


def addition(first, second):
    return [first[0]+second[0], first[1]+second[1]]


def multiplication(first, second):
    return [first[0]*second[0] - first[1]*second[1], first[0]*second[1] + first[1]*second[0]]


def absolute_value_square(number):
    return number[0]*number[0] + number[1]*number[1]


def round_to_decimal(number):
    return [round(number[0], 2), round(number[1], 2)]


def function(number, c):
    return round_to_decimal(addition(round_to_decimal(multiplication(number, number)), c))


test_input = [
    [-0.1, 0.75],
    [2.0, -0.3],
    [-0.5, 0.56],
    [-1.21, -0.32],
    [0.01, -0.64],
    [-0.52, 0.57],
    [-1.07, -0.2],
    [-1.04, -0.28],
    [0.33, 0.77],
    [0.26, -0.02]
]

test_output = [
    "3",
    "ESCAPES 2",
    "5",
    "8",
    "13",
    "5",
    "8",
    "21",
    "ESCAPES 6",
    "1"
]

for i in range(10):
    test_result = cycle_length(test_input[i][0], test_input[i][1])

    if test_result == test_output[i]:
        print("Test Case " + str(i) + ": Passed!")
    else:
        print("Test Case " + str(i) + ": Failed!")
        print("Expected output: " + test_output[i])
        print("Your output: " + test_result)

File: CPlusPlus/Contest 4/2021 - 2022/FibonacciMandelbrotIntermediate.cpp
/**
 *      ACSL 2021-2022 - Contest 4 - Fibonacci & Mandelbrot - Intermediate Division
 *
 */

#include <bits/stdc++.h>

using namespace std;

pair<double, double> addition(pair<double, double> A, pair<double, double> B)
{
    pair<double, double> result;
    result.first = A.first + B.first;
    result.second = A.second + B.second;

    return result;
}

pair<double, double> multiplication(pair<double, double> A, pair<double, double> B)
{
    pair<double, double> result;
    result.first = A.first*B.first - A.second*B.second;
    result.second = A.first*B.second + A.second*B.first;

    return result;
}

double absoluteValueSquare(pair<double, double> number)
{
    double result = number.first*number.first + number.second*number.second;

    return result;
}

pair<double, double> roundNumber(pair<double, double> number)
{
    pair<double, double> result;
    result.first = round(number.first*100) / 100.0;
    result.second = round(number.second*100) / 100.0;

    return result;
}

pair<double, double> transformation(pair<double, double> number, pair<double, double> C)
{
    pair<double, double> result = roundNumber(multiplication(number, number));
    result = roundNumber(addition(result, C)); 

    return result;
}

/*
 * Complete the 'cycleLength' function below.
 *
 * The function is expected to return a STRING.
 * The function accepts following parameters:
 *  1. FLOAT realPartC
 *  2. FLOAT imagPartC
 */
string cycleLength(float realPartC, float imagPartC)
{
    pair<double, double> C = {(double)realPartC, (double)imagPartC};
    pair<double, double> number = {0.0, 0.0};

    vector<pair<double, double>> numberList;
    numberList.push_back(number);

    string result = "";
    int n = 0;

    while (n <= 500)
    {
        n++;
        number = transformation(number, C);

        if (absoluteValueSquare(number) > 16)
        {
            result = "ESCAPES " + to_string(n);
            break;
        }

        bool found = false;
        for (int i = 0; i < n; i++)
        {
            if (numberList[i].first == number.first && numberList[i].second == number.second)
            {
                found = true;
                result = to_string(n - i);
                break;
            }
        }

        if (found)
        {
            break;
        }

        numberList.push_back(number);
    }

    return result;
}

int main()
{
    float input[10][2] = 
    {
        {-0.1f, 0.75f}, {2.0f, -0.3f}, {-0.5f, 0.56f}, {-1.21f, -0.32f}, {0.01f, -0.64f},
        {-0.52f, 0.57f}, {-1.07f, -0.2f}, {-1.04f, -0.28f}, {0.33f, 0.77f}, {0.26f, -0.02f}
    };

    string output[10] = 
    {
        "3", "ESCAPES 2", "5", "8", "13",
        "5", "8", "21", "ESCAPES 6", "1"
    };

    for (int i = 0; i < 10; i++)
    {
        string result = cycleLength(input[i][0], input[i][1]);

        if (output[i].compare(result) == 0) 
        {
            cout << "Test Case " << i << ": Passed!" << endl;
        }
        else 
        {
            cout << "Test Case " << i << ": Failed!" << endl;
            cout << "Expected output: " << output[i] << endl;
            cout << "Your output: " << result << endl;
        }
    }

    return 0;
}